plugin.audio.somafm/resources/lib/addon.py

```diff
@@ -0,0 +1,191 @@
+import urllib.parse
+import xml.etree.ElementTree as ET
+import time
+
+import xbmcaddon
+from xbmcgui import ListItem
+import xbmcplugin
+import xbmcgui
+import xbmc
+from xbmcplugin import SORT_METHOD_LISTENERS, SORT_METHOD_UNSORTED, SORT_METHOD_GENRE
+import xbmcvfs
+
+from resources.lib.channel import Channel
+
+CHANNELS_FILE_NAME = "channels.xml"
+
+__addon__ = "SomaFM"
+__addonid__ = "plugin.audio.somafm"
+__version__ = "2.0.0"
+
+__ms_per_day__ = 24 * 60 * 60 * 1000
+
+
+def log(msg):
+    xbmc.log(str(msg), level=xbmc.LOGDEBUG)
+
+
+log(sys.argv)
+
+rootURL = "https://somafm.com/"
+tempdir = xbmcvfs.translatePath("special://home/userdata/addon_data/%s" % __addonid__)
+xbmcvfs.mkdirs(tempdir)
+
+LOCAL_CHANNELS_FILE_PATH = os.path.join(tempdir, CHANNELS_FILE_NAME)
+
+try:
+    plugin_url = sys.argv[0]
+    handle = int(sys.argv[1])
+    query = sys.argv[2]
+except Exception as e:
+    xbmc.log("Initialization Failed: " + str(e), level=xbmc.LOGERROR)
+    plugin_url = "plugin://" + __addonid__
+    handle = 0
+    query = ""
+
+
+def fetch_remote_channel_data():
+    with urllib.request.urlopen(rootURL + CHANNELS_FILE_NAME) as response:
+        channel_data = response.read()
+        with open(LOCAL_CHANNELS_FILE_PATH, 'w') as local_channels_xml:
+            local_channels_xml.write(channel_data.decode('utf-8'))
+        return channel_data
+
+
+def fetch_local_channel_data():
+    with open(LOCAL_CHANNELS_FILE_PATH) as local_channels_file:
+        return local_channels_file.read()
+
+
+def fetch_cached_channel_data():
+    if os.path.getmtime(LOCAL_CHANNELS_FILE_PATH) + cache_ttl_in_ms() > time.time():
+        return fetch_local_channel_data()
+    # don't delete the cached file so we can still use it as a fallback
+    # if something goes wrong fetching the channel data from server
+
+
+def fetch_channel_data(*strategies):
+    for strategy in strategies:
+        try:
+            result = strategy()
+            if result is not None:
+                return result
+        except Exception as e:
+            xbmc.log("fetch_channel_data Failed: " + str(e), level=xbmc.LOGERROR)
+            pass
+
+
+def build_directory():
+    channel_data = fetch_channel_data(fetch_cached_channel_data, fetch_remote_channel_data, fetch_local_channel_data)
+    xml_data = ET.fromstring(channel_data)
+
+    stations = xml_data.findall(".//channel")
+    for station in stations:
+        channel = Channel(handle, tempdir, station)
+        li = xbmcgui.ListItem(
+            channel.get_simple_element('title'),
+            channel.get_simple_element('description'),
+            plugin_url + channel.getid())
+
+        li.setArt({
+            "icon": channel.geticon(),
+            "thumb": channel.getthumbnail(),
+            "fanart": xbmcvfs.translatePath("special://home/addons/%s/fanart.jpg" % __addonid__)
+        })
+
+        li.setProperty("IsPlayable", "true")
+
+        for element, info in [('listeners', 'listeners'),
```

Comment:
heads up setInfo with dicts is deprecated in Nexus and will be removed in further versions, we have now dedicated classes for info

Reply:
Could I trouble you for a link to an example of how it should be done going forward? I am only able to find dictionary-based approaches documented.

Reply:
Maybe the pr is the best source of info: https://github.com/xbmc/xbmc/pull/19459#issuecomment-806450382

Reply:
Also doxygen (bit broken at the moment): https://codedocs.xyz/xbmc/xbmc/group__python___info_tag_music.html